infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/PathBasePrefixProvider.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.jwtproxy;
+
+import static java.lang.String.format;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.DefaultHostExternalServiceExposureStrategy.DEFAULT_HOST_STRATEGY;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.IngressServiceExposureStrategyProvider.STRATEGY_PROPERTY;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.MultiHostExternalServiceExposureStrategy.MULTI_HOST_STRATEGY;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.SingleHostExternalServiceExposureStrategy.SINGLE_HOST_STRATEGY;
+
+import com.google.inject.Inject;
+import java.util.function.Function;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+
+/**
+ * This is a helper class to {@link JwtProxyProvisioner} that helps it figure out the prefix on the
+ * "pathBase" on which it should be exposed. This is used both as a prefix on the pathBase and as
+ * the path set on the authentication cookie.
+ *
+ * <p>The prefix is based on the workspace ID so that we can have separate authentication cookies
+ * per workspace.
+ */
+@Singleton
+public class PathBasePrefixProvider {
+
+  private final Function<RuntimeIdentity, String> extractor;
+
+  @Inject
+  public PathBasePrefixProvider(@Named(STRATEGY_PROPERTY) String serverStrategy) {
+    switch (serverStrategy) {
+      case DEFAULT_HOST_STRATEGY:
+      case SINGLE_HOST_STRATEGY:
+        extractor = RuntimeIdentity::getWorkspaceId;
```

Comment:
Do you happen to know how K8s behalves if you try to create two ingresses with the same http path rule?

I just wondering it using common single cookies path won't lead to security issue...
Like user creates a workspace, opens it, and as result has access token in a cookie that is configured with `/workspace123` path.
The attacker knows workspace id, creates his own ingress with `/workspace123/show-me-your-token` path and service/pod behind just to ready user token.
Then it's pretty clear - attacker shares with the user a page which sends a request to `/workspace123/show-me-your-token`.

As far as I understand, previously it was not possible.
If you set his own cookies path for each server, like `/workspace123/server-123` then it won't be possible as well (actually depending of k8s behavior when two ingresses with the same path exist) because `/workspace123/show-me-your-token` will not reuse already set access token in cookies, and `loader.js` will reject tries to steal token with error: `/workspace123/show-me-your-token` is not a part of `workspace123`.

I may be wrong, it's just thought that came to my mind but please analyze if it makes sense.